httpbp/client_middlewares.go

```diff
@@ -0,0 +1,199 @@
+package httpbp
+
+import (
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		MonitorClient(config),
+		Retries(),
+		CircuitBreaker(config),
+	}
+
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+//
+// Adding this middleware means that both, non-nil response and non-nil error,
+// can be returned. If configured the caller needs to call
+// httpbp.DrainAndClose(resp.Body) to ensure the underlying TCP connection can
+// be re-used.
```

Comment:
it's probably worth adding an example here, something like:

```
// Instead of doing this:
//
//     // bad example, DO NOT USE
//     resp, err := client.Do(req)
//     if err != nil {
//       // TODO: handle error
//       return err
//     }
//     defer httpbp.DrainAndClose(resp.Body)
//     // TODO: handle resp
//
// Do this:
//
//     resp, err := client.Do(req)
//     if resp != nil {
//       defer httpbp.DrainAndClose(resp.Body)
//     }
//     if err != nil {
//       // TODO: handle error
//       return err
//     }
//     // TODO: handle resp
```

Reply:
I'll add that. Should I remove `TODO: handle error` since returning the error is a way of delegating the error handler to the caller or do we keep it in there to remind people to handle errors eventually?

Reply:
Examples should be done in the _test.go with an Example function; this will show up in godoc but it will also be compiled to ensure it won't drift.

Reply:
Why not
```
resp, err := client.Do(req)
defer httpbp.DrainAndClose(resp)
if err != nil {
  // TODO: handle error
}
```

This lets you check for both nil responses and bodies inside DrainAndClose.

Reply:
since we are going to handle the body in this middleware, that example in the comments is no longer needed.

Reply:
You can put the "don't do" in the example, something like

```
if false {
  // Don't do this:
}

// Do this:
```